src/mbgl/style/expression/match.cpp

```diff
@@ -0,0 +1,33 @@
+#include <mbgl/style/expression/match.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+template<> Result<std::string> TypedMatch<std::string>::evaluateInput(const EvaluationParameters& params) const {
+    const auto& inputValue = input->evaluate<std::string>(params);
+    if (!inputValue) {
+        return inputValue.error();
+    }
+    return *inputValue;
+}
+
+template<> Result<int64_t> TypedMatch<int64_t>::evaluateInput(const EvaluationParameters& params) const {
+    const auto& inputValue = input->evaluate<float>(params);
+    if (!inputValue) {
+        return inputValue.error();
+    }
+    int64_t rounded = ceilf(*inputValue);
+    if (*inputValue == rounded) {
+        return rounded;
+    } else {
+        return EvaluationError {
```

Comment:
Yeah, you're right; since we don't distinguish among numeric types, this shouldn't be treated as a type error